From 5548f687a583638605b1b2d794ec6597ded68108 Mon Sep 17 00:00:00 2001
From: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Date: Tue, 13 Nov 2018 11:01:36 +0200
Subject: [PATCH 5119/5242] MLK-20117 drm/imx/dcss: fix color issue when
 Adobe_ARGB gamut is used

commit  34e6d1babc36dd78112328307391403bff4bdcdf from
https://source.codeaurora.org/external/imx/linux-imx.git

Adobe ARGB gamut was selected even if the output pipe pixel encoding was
YUV. That produced a pink tint on the screen.

This patch will make sure the Adobe ARGB gamut is selected only when the
output pipe pixel encoding is RGB.

Signed-off-by: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/dcss/dcss-crtc.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/imx/dcss/dcss-crtc.c b/drivers/gpu/drm/imx/dcss/dcss-crtc.c
index 4962eb1..9cf8ef6 100644
--- a/drivers/gpu/drm/imx/dcss/dcss-crtc.c
+++ b/drivers/gpu/drm/imx/dcss/dcss-crtc.c
@@ -191,13 +191,16 @@ void dcss_crtc_setup_opipe(struct drm_crtc *crtc, struct drm_connector *conn,
 	if ((colorimetry & BIT(HDMI_EXTENDED_COLORIMETRY_BT2020)) ||
 	    (colorimetry & BIT(HDMI_EXTENDED_COLORIMETRY_BT2020_CONST_LUM)))
 		dcss_crtc->opipe_g = G_REC2020;
-	else if (colorimetry & BIT(HDMI_EXTENDED_COLORIMETRY_ADOBE_RGB))
+	else if ((colorimetry & BIT(HDMI_EXTENDED_COLORIMETRY_ADOBE_RGB)) &&
+		 !(crtc->state->adjusted_mode.private_flags & YUV_MODE))
 		dcss_crtc->opipe_g = G_ADOBE_ARGB;
 	else
 		dcss_crtc->opipe_g = G_REC709;
 
 	if ((eotf & (1 << 2)) && dcss_crtc->opipe_g == G_REC2020)
 		dcss_crtc->opipe_nl = NL_REC2084;
+	else if (dcss_crtc->opipe_g == G_ADOBE_ARGB)
+		dcss_crtc->opipe_nl = NL_SRGB;
 	else
 		dcss_crtc->opipe_nl = NL_REC709;
 
-- 
1.7.9.5

